### PR TITLE
DEX-521 Wrong cancel

### PR DESCRIPTION
**Pull request by vsuharnikov**
* Added CreateExchangeTransactionActor. It listens for events from AddressActor and creates a transactions. An event means that AddressActor observed OrderExecuted;
* OrderBookActor no longer creates transactions;
* AddressActor shows more information about an order canceled during balance checks;
* Added an integration test;

### CODE REVIEW

**Comment by vsuharnikov** [comment]
Commented tests, because we depend on https://github.com/wavesplatform/Waves/pull/2770

File: dex/src/main/scala/com/wavesplatform/dex/market/CreateExchangeTransactionActor.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import com.wavesplatform.account.Address
+import com.wavesplatform.dex.market.CreateExchangeTransactionActor.OrderExecutedObserved
+import com.wavesplatform.dex.model.Events.{ExchangeTransactionCreated, OrderExecuted}
+import com.wavesplatform.dex.model.ExchangeTransactionCreator.CreateTransaction
+import com.wavesplatform.utils.ScorexLogging
+import play.api.libs.json.Json
+
+import scala.collection.mutable
+
+class CreateExchangeTransactionActor(createTransaction: CreateTransaction) extends Actor with ScorexLogging {
```
Comment (DenisSerebryansky) [inline dex/src/main/scala/com/wavesplatform/dex/market/CreateExchangeTransactionActor.scala]: Can you please add some brief explanation for the main purpose of this actor?
I mean the idea about awaiting confirmation from both address actors (or one) can be wrote down here in a short doc

File: dex/src/main/scala/com/wavesplatform/dex/market/OrderBookActor.scala [hunk 2]
```diff
@@ -127,21 +125,14 @@ class OrderBookActor(owner: ActorRef,
...
-            case Right(tx) =>
-              log.info(s"Created transaction: $tx")
-              context.system.eventStream.publish(ExchangeTransactionCreated(tx))
-            case Left(ex) =>
-              log.warn(s"""Can't create tx: $ex
-                   |o1: (amount=${submitted.amount}, fee=${submitted.fee}): ${Json.prettyPrint(submitted.order.json())}
-                   |o2: (amount=${counter.amount}, fee=${counter.fee}): ${Json.prettyPrint(counter.order.json())}""".stripMargin)
-          }
-      }
+          s"OrderExecuted(s=${submitted.order.idStr()}, c=${counter.order.idStr()}, amount=${oe.executedAmount})"
```
Comment (DenisSerebryansky) [inline dex/src/main/scala/com/wavesplatform/dex/market/OrderBookActor.scala]: Could it be oneliners?

Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/market/OrderBookActor.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/market/CreateExchangeTransactionActor.scala (hunk 1 above)
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/market/CreateExchangeTransactionActor.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala [hunk 3]
```diff
@@ -374,14 +377,14 @@ class AddressActor(owner: Address,
...
-      case (Some(curr), (assetId, amount)) =>
+  private def remove(from: SpendableBalance, xs: SpendableBalance): Either[(Long, Asset), SpendableBalance] =
+    xs.foldLeft[Either[(Long, Asset), SpendableBalance]](Right(from)) {
+      case (r @ Left(_), _) => r
+      case (curr, (_, 0))   => curr
+      case (Right(curr), (assetId, amount)) =>
         val updatedAmount = curr.getOrElse(assetId, 0L) - amount
-        if (updatedAmount < 0) None
-        else Some(curr.updated(assetId, updatedAmount))
+        if (updatedAmount < 0) Left((updatedAmount, assetId))
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala]: Done

**Pull request merged**